skimage/segmentation/_join.py

```diff
@@ -333,8 +332,24 @@ def __getitem__(self, index):
         return out
 
     def __setitem__(self, indices, values):
-        if self._array is None:
-            self._array = self.__array__()
-        self._array[indices] = values
-        self.in_values = np.flatnonzero(self._array)
-        self.out_values = self._array[self.in_values]
+        if np.isscalar(indices):
+            indices = np.atleast_1d(indices)
+        elif isinstance(indices, slice):
+            start = indices.start or 0  # treat None or 0 the same way
+            stop = (indices.stop
+                    if indices.stop is not None
+                    else len(self))
+            step = indices.step
+            indices = np.arange(start, stop, step)
+        if indices.dtype == bool:
+            indices = np.flatnonzero(indices)
+
+        if np.isscalar(values) or len(values) == 1:
+            values = np.repeat(values, len(indices))
+
+        for ind, val in zip(indices, values):
+            if ind in self.in_values:
+                self.out_values[self.in_values == ind] = val
+            else:
+                self.in_values = np.append(self.in_values, ind)
+                self.out_values = np.append(self.out_values, val)
```

Comment:
@rfezzani the problem is that, if you look at the *two* cases out of 240 that used `setitem`, they used it for almost the entire array. Specifically, they did `fw[1:] += new_fw`. So, you end up allocating an equally huge array, and you do so very very slowly, using quadratic growth of the array. So imho this should be scrapped...

Reply:
I also though about this case and I will come with a proposal for implementing the `__iadd__` operator, since in this case we loose the sparseness of the `ArrayMap`